### ISSUE TITLE
dict-comprehension on mapping from a dataclass of sequence containing literal remove literal information

### ISSUE THREAD

**Issue by Holt59** [microsoft/pyright#8052]
**Describe the bug**

I have a variable that is a `Mapping` where values are `Sequence` of literals. When re-assigning the mapping by copying the sequence, the literal information is lost.

**Code or Screenshots**

I cannot provide a completely reproducible example, but this is what I can give

```python
from collections.abc import Mapping
from typing import Literal, Sequence


class Data:
    mapping: Mapping[int, Sequence[Literal["a", "b"]]]


def fn(
    data: Data,
    convert: bool = True,
):
    mapping = data.mapping
    reveal_type(mapping)
    if convert:
        mapping = {k: [t for t in mapping[k]] for k in mapping}
        reveal_type(mapping)
```

Output:

```
test.py:14:17 - information: Type of "mapping" is "Mapping[int, Sequence[Literal['a', 'b']]]"
test.py:17:21 - information: Type of "mapping" is "dict[int, list[str]]
```

As you can see, the second `reveal_type` has removed the `Literal` information.

**Note:** This does not occur if I pass the `mapping` variable directly to `fn` (i.e., replace `data: Data` by `mapping: Mapping[int, Sequence[Literal["a", "b"]]]`).

**VS Code extension or command-line**

This is most likely a regression (if this is not the expected result) from 1.1.364 to 1.1.365:

- Problem does not appear in VS code currently (Pylance v2024.5.1)
- Problem was not occurring with CLI 1.1.364
- Problem is occurring with CLI 1.1.365

**Comment by erictraut** [comment]
Yes, this behavior is intended. Pyright's default inference behavior for list expressions does not retain literals. After all, you probably wouldn't expect `[1]` to evaluate to `list[Literal[1]]`.

If you want to retain literal types, you will need to provide an [bidirectional inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types) context, such as an explicit type declaration for the `mapping` variable.

**Comment by Holt59** [comment]
> Yes, this behavior is intended. Pyright's default inference behavior for list expressions does not retain literals. After all, you probably wouldn't expect `[1]` to evaluate to `list[Literal[1]]`.
> 
> If you want to retain literal types, you will need to provide an [bidirectional inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types) context, such as an explicit type declaration for the `mapping` variable.

Is there a way to force explicit type in the assignment? Having to type very long mapping declaration for all these cases feel very wrong when the assigned value is already fully typed. I'd expect the type to be maintain.

**Comment by Holt59** [comment]
And is there a place where the distinction between explicit / non-explicit type behavior is documented? I always assumed that implicit typing would provide the same behavior as explicit one, but that does not seem to be the case.

**Comment by erictraut** [comment]
Pyright's default type inference rules are documented [here](https://microsoft.github.io/pyright/#/type-inference). There are often multiple "correct" types for a given expression, so regardless of how a type checker's default inference rules are defined, cases where it is necessary to override those inference rules. See [this documentation](https://microsoft.github.io/pyright/#/type-inference?id=ambiguous-type-inference) for an explanation.
